File: c++/2023/solutions/day01-1.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <cctype>

int main() {
	std::ifstream input{ "../inputs/2023/day01-input.txt" };
	if (!input.is_open()) std::cout << "file did not open\n";
	
	std::string s;
	int sum{};
	while (std::getline(input, s)) {
		std::string num;
		char first = '/';
		char last = '/';
		for (char c : s) {
			if (std::isdigit(c)) {
				last = c;
				if (first == '/') first = c;
			}
		}
		int n = (first-'0') * 10 + (last-'0');
		sum += n;
	}
	std::cout << sum << '\n';
}

File: haskell/2023/solutions/day01-1.hs
import Data.Char
import Data.List as L

firstDigit :: String -> Char
firstDigit x = maybe '0' id (L.find isDigit x)

lastDigit :: String -> Char
lastDigit x = firstDigit $ L.reverse x

calibration :: String -> Int
calibration x = read [firstDigit x, lastDigit x]

solution :: String -> String
solution x = show $ sum $ L.map calibration (L.lines x)

main = do
       file <- readFile "../inputs/2023/day01-input.txt"
       putStrLn $ solution file
       